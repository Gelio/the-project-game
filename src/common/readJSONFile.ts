import { readFile as fsReadFile } from 'fs';

type ReadFileFn = (
  path: string,
  options: { encoding: string; flag?: string } | string,
  callback: (err: NodeJS.ErrnoException, data: string) => void
) => void;

export function readJSONFile(filePath: string, readFile: ReadFileFn = fsReadFile): Promise<object> {
  return new Promise((resolve, reject) => {
    readFile(filePath, { encoding: 'utf8' }, (readError, data) => {
      if (readError) {
        reject(readError);

        return;
      }

      try {
        resolve(JSON.parse(data));
      } catch (parseError) {
        reject(parseError);
      }
    });
  });
}
