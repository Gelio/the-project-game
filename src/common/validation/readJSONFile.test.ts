import { readJSONFile } from './readJSONFile';

describe('readJSONFile', () => {
  it('should reject with an error when reading the file fails', () => {
    function readFile(_path: string, _options: any, callback: Function) {
      callback(new Error('test'));
    }

    expect(readJSONFile('path', readFile)).rejects.toMatchSnapshot();
  });

  it('should reject with an error when parsing the file fails', () => {
    function readFile(_path: string, _options: any, callback: Function) {
      callback(null, 'not valid JSON');
    }

    expect(readJSONFile('path', readFile)).rejects.toMatchSnapshot();
  });

  it('should resolve with the parsed object', () => {
    const obj = {
      test: 'foo'
    };

    function readFile(_path: string, _options: any, callback: Function) {
      callback(null, JSON.stringify(obj));
    }

    expect(readJSONFile('path', readFile)).resolves.toEqual(obj);
  });

  it('should read the file from the specified path', () => {
    const readFile = jest.fn();
    const filePath = 'foo';

    readJSONFile(filePath, readFile);

    expect(readFile.mock.calls[0][0]).toBe(filePath);
  });
});
