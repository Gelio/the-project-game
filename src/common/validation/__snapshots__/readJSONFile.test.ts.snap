// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readJSONFile should reject with an error when parsing the file fails 1`] = `[SyntaxError: Unexpected token o in JSON at position 1]`;

exports[`readJSONFile should reject with an error when reading the file fails 1`] = `[Error: test]`;
